kaldb/src/main/java/com/slack/kaldb/clusterManager/ClusterHpaMetricService.java

```diff
@@ -0,0 +1,172 @@
+  protected final Map<String, Instant> cacheScalingLock = new ConcurrentHashMap<>();
+  protected static final String CACHE_HPA_METRIC_NAME = "hpa_cache_demand_factor_%s";
+
+  public ClusterHpaMetricService(
+      ReplicaMetadataStore replicaMetadataStore,
+      CacheSlotMetadataStore cacheSlotMetadataStore,
+      HpaMetricMetadataStore hpaMetricMetadataStore) {
+    this.replicaMetadataStore = replicaMetadataStore;
+    this.cacheSlotMetadataStore = cacheSlotMetadataStore;
+    this.hpaMetricMetadataStore = hpaMetricMetadataStore;
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(Duration.ofSeconds(15), Duration.ofSeconds(30));
+  }
+
+  @Override
+  protected void runOneIteration() {
+    LOG.info("Running ClusterHpaMetricService");
+    try {
+      publishCacheHpaMetrics();
+    } catch (Exception e) {
+      LOG.error("Error running ClusterHpaMetricService", e);
+    }
+  }
+
+  /**
+   * Calculates and publishes HPA scaling metric(s) to Zookeeper for the Cache nodes. This makes use
+   * of the Kubernetes HPA formula, which is:
+   *
+   * <p>desiredReplicas = ceil[currentReplicas * ( currentMetricValue / desiredMetricValue )]
+   *
+   * <p>Using this formula, if we inform the user what value to set for the desiredMetricValue, we
+   * can vary the currentMetricValue to control the scale of the cluster. This is accomplished by
+   * producing a metric (hpa_cache_demand_factor_REPLICASET) that targets a value of 1.0. As long as
+   * the HPA is configured to target a value of 1.0 this will result in an optimized cluster size
+   * based on cache slots vs replicas, plus a small buffer.
+   *
+   * <pre>
+   * metrics:
+   *   - type: Pods
+   *     pods:
+   *       metric:
+   *         name: hpa_cache_demand_factor_REPLICASET
+   *       target:
+   *         type: AverageValue
+   *         averageValue: 1.0
+   * </pre>
+   */
+  private void publishCacheHpaMetrics() {
+    Set<String> replicaSets =
+        replicaMetadataStore.listSync().stream()
+            .map(ReplicaMetadata::getReplicaSet)
+            .collect(Collectors.toSet());
+
+    for (String replicaSet : replicaSets) {
+      long totalCacheSlotCapacity =
+          cacheSlotMetadataStore.listSync().stream()
+              .filter(cacheSlotMetadata -> cacheSlotMetadata.replicaSet.equals(replicaSet))
+              .count();
+      long totalReplicaDemand =
+          replicaMetadataStore.listSync().stream()
+              .filter(replicaMetadata -> replicaMetadata.getReplicaSet().equals(replicaSet))
+              .count();
+      double rawDemandFactor =
+          (double) (totalReplicaDemand + CACHE_SLOT_OVER_PROVISION) / (totalCacheSlotCapacity + 1);
+      double demandFactor = (double) Math.round(rawDemandFactor * 100) / 100;
+      LOG.info(
+          "Cache autoscaler for replicaSet '{}' calculated a demandFactor of '{}' - totalReplicaDemand: '{}', cacheSlotOverProvision: '{}', totalCacheSlotCapacity: '{}'",
```

Comment:
will this log line be too noisy? The scheduler runs every 30s and two replcasets.

Reply:
This is the easiest way to troubleshoot the behavior and determine what's going on with the logic - we may be able to reduce it in the future, but this would only be four lines per min which is negligible for our overall cluster.